src/compiler/crystal/codegen/debug.cr

```diff
@@ -110,40 +125,182 @@ module Crystal
       end
 
       size = @program.target_machine.data_layout.size_in_bits(struct_type)
-      debug_type = di_builder.create_struct_type(nil, type.to_s, nil, 1, size, size, LLVM::DIFlags::Zero, nil, di_builder.get_or_create_type_array(element_types))
+      debug_type = di_builder.create_struct_type(nil, type_name, nil, 1, size, size, LLVM::DIFlags::Zero, nil, di_builder.get_or_create_type_array(element_types))
       unless type.struct?
-        debug_type = di_builder.create_pointer_type(debug_type, 8u64 * llvm_typer.pointer_size, 8u64 * llvm_typer.pointer_size, type.to_s)
+        debug_type = di_builder.create_pointer_type(debug_type, 8u64 * llvm_typer.pointer_size, 8u64 * llvm_typer.pointer_size, type_name)
       end
       di_builder.replace_temporary(tmp_debug_type, debug_type)
       debug_type
     end
 
-    def create_debug_type(type : PointerInstanceType)
+    def create_debug_type(type : (PointerInstanceType | Pointer(T)), type_name : String? = type.to_s)
       element_type = get_debug_type(type.element_type)
       return unless element_type
-      di_builder.create_pointer_type(element_type, 8u64 * llvm_typer.pointer_size, 8u64 * llvm_typer.pointer_size, type.to_s)
+      di_builder.create_pointer_type(element_type, 8u64 * llvm_typer.pointer_size, 8u64 * llvm_typer.pointer_size, type_name)
+    end
+
+    def create_debug_type(type : MixedUnionType, type_name : String? = type.to_s)
+      element_types = [] of LibLLVMExt::Metadata
+      struct_type = llvm_type(type)
+      struct_type_size = @program.target_machine.data_layout.size_in_bits(struct_type)
+      is_struct = struct_type.struct_element_types.size == 1
+
+      tmp_debug_type = di_builder.create_replaceable_composite_type(nil, type_name, nil, 1, llvm_context)
+      debug_type_cache[type_name] = tmp_debug_type
+
+      type.expand_union_types.each do |ivar_type|
+        next if ivar_type.is_a?(NilType) || ivar_type.to_s == "Nil"
```

Comment:
Is `|| ivar_type.to_s == "Nil"` condition rly required?

Reply:
Removed as it was redundant.